server/src/main/java/org/elasticsearch/index/mapper/flattened/FlattenedFieldSyntheticWriterHelper.java

```diff
@@ -0,0 +1,308 @@
+ *             "items": ["102202929", "290092911"],
+ *         },
+ *         "to": "Atlanta"
+ *     }
+ * }`
+ *
+ */
+class FlattenedFieldSyntheticWriterHelper {
+
+    private static class Prefix {
+        final List<String> prefix;
+
+        Prefix() {
+            this.prefix = Collections.emptyList();
+        }
+
+        Prefix(final String key) {
+            final String[] keyAsArray = key.split("\\.");
+            this.prefix = new ArrayList<>(Arrays.asList(keyAsArray).subList(0, keyAsArray.length - 1));
+        }
+
+        private Prefix(final List<String> prefix) {
+            this.prefix = prefix;
+        }
+
+        private Prefix shared(final Prefix other) {
+            return shared(this.prefix, other.prefix);
+        }
+
+        private static Prefix shared(final List<String> curr, final List<String> next) {
+            final List<String> shared = new ArrayList<>();
+            for (int i = 0; i < Math.min(curr.size(), next.size()); i++) {
+                if (curr.get(i).equals(next.get(i))) {
+                    shared.add(curr.get(i));
+                }
+            }
+
+            return new Prefix(shared);
+        }
+
+        private Prefix diff(final Prefix other) {
+            return diff(this.prefix, other.prefix);
+        }
+
+        private static Prefix diff(final List<String> a, final List<String> b) {
+            if (a.size() > b.size()) {
+                return diff(b, a);
+            }
+            final List<String> diff = new ArrayList<>();
+            if (a.isEmpty()) {
+                diff.addAll(b);
+                return new Prefix(diff);
+            }
+            int i = 0;
+            for (; i < a.size(); i++) {
+                if (a.get(i).equals(b.get(i)) == false) {
+                    break;
+                }
+            }
+            for (; i < b.size(); i++) {
+                diff.add(b.get(i));
+            }
+            return new Prefix(diff);
+        }
+
+        private Prefix reverse() {
+            final Prefix p = new Prefix(this.prefix);
+            Collections.reverse(p.prefix);
+            return p;
+        }
+
+        @Override
+        public int hashCode() {
+            return Objects.hash(this.prefix);
+        }
+
+        @Override
+        public boolean equals(Object obj) {
+            if (obj == null) {
+                return false;
+            }
+            if (getClass() != obj.getClass()) {
+                return false;
+            }
+            Prefix other = (Prefix) obj;
+            return Objects.equals(this.prefix, other.prefix);
+        }
+    }
+
+    private static class Suffix {
+        final List<String> suffix;
+
+        Suffix() {
+            this.suffix = Collections.emptyList();
+        }
+
+        Suffix(final String key) {
+            final String[] keyAsArray = key.split("\\.");
+            this.suffix = new ArrayList<>(Arrays.asList(keyAsArray).subList(keyAsArray.length - 1, keyAsArray.length));
```

Comment:
Can `List.of(keyAsArray).subList(keyAsArray.length - 1, keyAsArray.length))` be used here?